Source/JavaScriptCore/runtime/PropertyTable.h

```diff
@@ -238,182 +213,240 @@ class PropertyTable final : public JSCell {
+
+    bool isCompact() const { return m_indexVector & isCompactFlag; }
+
+    template<typename Functor>
+    void forEachPropertyMutable(const Functor&);
+
+    // The table of values lies after the hash index.
+    CompactPropertyTableEntry* tableFromIndexVector(uint8_t* index)
+    {
+        return reinterpret_cast_ptr<CompactPropertyTableEntry*>(index + m_indexSize);
+    }
+    const CompactPropertyTableEntry* tableFromIndexVector(const uint8_t* index) const
+    {
+        return reinterpret_cast_ptr<const CompactPropertyTableEntry*>(index + m_indexSize);
+    }
+    PropertyTableEntry* tableFromIndexVector(uint32_t* index)
+    {
+        return reinterpret_cast_ptr<PropertyTableEntry*>(index + m_indexSize);
+    }
+    const PropertyTableEntry* tableFromIndexVector(const uint32_t* index) const
+    {
+        return reinterpret_cast_ptr<const PropertyTableEntry*>(index + m_indexSize);
+    }
+
+    CompactPropertyTableEntry* tableEndFromIndexVector(uint8_t* index)
+    {
+        return tableFromIndexVector(index) + usedCount();
+    }
+    const CompactPropertyTableEntry* tableEndFromIndexVector(const uint8_t* index) const
+    {
+        return tableFromIndexVector(index) + usedCount();
+    }
+    PropertyTableEntry* tableEndFromIndexVector(uint32_t* index)
+    {
+        return tableFromIndexVector(index) + usedCount();
+    }
+    const PropertyTableEntry* tableEndFromIndexVector(const uint32_t* index) const
+    {
+        return tableFromIndexVector(index) + usedCount();
+    }
+
+    static constexpr uintptr_t isCompactFlag = 0x1;
+    static constexpr uintptr_t indexVectorMask = ~isCompactFlag;
 
     unsigned m_indexSize;
     unsigned m_indexMask;
-    unsigned* m_index;
+    uintptr_t m_indexVector;
     unsigned m_keyCount;
     unsigned m_deletedCount;
     std::unique_ptr<Vector<PropertyOffset>> m_deletedOffsets;
 
     static constexpr unsigned MinimumTableSize = 16;
+    static_assert(MinimumTableSize >= 16, "compact index is uint8_t and we should keep 16 byte aligned entries after this array");
 };
 
-inline PropertyTable::iterator PropertyTable::begin()
-{
-    auto* tableEnd = this->tableEnd();
-    return iterator(skipDeletedEntries(table(), tableEnd), tableEnd);
-}
-
-inline PropertyTable::iterator PropertyTable::end()
-{
-    auto* tableEnd = this->tableEnd();
-    return iterator(tableEnd, tableEnd);
-}
-
-inline PropertyTable::const_iterator PropertyTable::begin() const
-{
-    auto* tableEnd = this->tableEnd();
-    return const_iterator(skipDeletedEntries(table(), tableEnd), tableEnd);
-}
-
-inline PropertyTable::const_iterator PropertyTable::end() const
-{
-    auto* tableEnd = this->tableEnd();
-    return const_iterator(tableEnd, tableEnd);
-}
-
-inline PropertyTable::find_iterator PropertyTable::find(const KeyType& key)
+template<typename Index, typename Entry>
+std::tuple<unsigned, unsigned, PropertyOffset, uint8_t> PropertyTable::findImpl(const Index* indexVector, const Entry* table, const KeyType& key)
```

Comment:
nit: Can we make a simple struct here instead of tuple. tuple with 4 elements is hard to follow compared to

```
struct FindResult {
    unsigned entryIndex;
    unsigned index;
    PropertyOffset offset;
    unsigned attributes;
};

Reply:
Done